semgrep-core/src/engine/Selector.ml

```diff
@@ -0,0 +1,130 @@
+module R = Rule
+module AR = Augmented_range
+module G = AST_generic
+module M = Metavariable
+module PM = Pattern_match
+module RP = Report
+
+type selector = {
+  mvar : M.mvar;
+  pid : int;
+  pstr : string;
+  lazy_matches : RP.times RP.match_result Lazy.t;
+}
+
+type sformula =
+  | Leaf of R.leaf
+  | And of (selector option * sformula list)
+  | Or of sformula list
+  (* There are restrictions on where a Not can appear in a formula. It
+   * should always be inside an And to be intersected with "positive" formula.
+   *)
+  | Not of sformula
+
+(*****************************************************************************)
+(* Selecting methods *)
+(*****************************************************************************)
+
+(* this will be adjusted later in range_to_pattern_match_adjusted *)
+let fake_rule_id (id, str) =
+  { PM.id = string_of_int id; pattern_string = str; message = "" }
+
+let match_selector (sel_opt : selector option) : AR.ranges =
+  let get_match (x : RP.times RP.match_result) = x.matches in
+  match sel_opt with
+  | None -> failwith "empty And; no positive terms in And"
+  | Some selector ->
+      List.map AR.match_result_to_range
+        (get_match (Lazy.force selector.lazy_matches))
+
+let select_from_ranges file (sel_opt : selector option) (ranges : AR.ranges) :
+    AR.ranges =
+  let pattern_match_from_binding selector (mvar, mval) =
+    {
+      PM.rule_id = fake_rule_id (selector.pid, selector.pstr);
+      PM.file;
+      PM.range_loc = Visitor_AST.range_of_any (M.mvalue_to_any mval);
+      PM.tokens = lazy (M.ii_of_mval mval);
+      PM.env = [ (mvar, mval) ];
+    }
+  in
+  let select_from_range range =
+    match sel_opt with
+    | None -> [ range ]
+    | Some selector -> (
+        match
+          List.find_opt
+            (fun (mvar, _mval) -> M.equal_mvar selector.mvar mvar)
+            range.AR.mvars
+        with
+        | None -> []
+        | Some binding ->
+            (* make a pattern match, then use AR.match_result_to_range *)
+            [
+              AR.match_result_to_range
+                (pattern_match_from_binding selector binding);
+            ] )
+  in
+  List.flatten (List.map select_from_range ranges)
+
+let selector_equal s1 s2 = s1.mvar = s2.mvar
+
+(*****************************************************************************)
+(* Converter *)
+(*****************************************************************************)
+
+let formula_to_sformula match_func formula =
+  let rec formula_to_sformula formula =
+    let selector_from_formula f =
```

Comment:
This function could be pull out of the `let rec formula_to_sformula`, would make it clear that `selector_from_formula` can be understood separately and does not depend on any of the arguments of `let rec formula_to_sformula`.

Reply:
Yep, @emjin it's usually better to avoid nested functions. In the end it makes the code a bit more complicated.